cores/arduino/stm32/analog.c

```diff
@@ -638,32 +790,161 @@ uint16_t adc_read_value(PinName pin)
   /*  For simplicity reasons, this example is just waiting till the end of the
       conversion, but application may perform other tasks while conversion
       operation is ongoing. */
-  if (HAL_ADC_PollForConversion(&AdcHandle, 10) != HAL_OK)
+  if (HAL_ADC_PollForConversion(AdcHandle, 10) != HAL_OK)
   {
     /* End Of Conversion flag not set on time */
     return 0;
   }
 
   /* Check if the continous conversion of regular channel is finished */
-  if ((HAL_ADC_GetState(&AdcHandle) & HAL_ADC_STATE_REG_EOC) == HAL_ADC_STATE_REG_EOC)
+  if ((HAL_ADC_GetState(AdcHandle) & HAL_ADC_STATE_REG_EOC) == HAL_ADC_STATE_REG_EOC)
   {
     /*##-5- Get the converted value of regular channel  ########################*/
-    uhADCxConvertedValue = HAL_ADC_GetValue(&AdcHandle);
+    uhADCxConvertedValue = HAL_ADC_GetValue(AdcHandle);
   }
 
-  if (HAL_ADC_Stop(&AdcHandle) != HAL_OK)
+  if (HAL_ADC_Stop(AdcHandle) != HAL_OK)
   {
     /* Stop Conversation Error */
     return 0;
   }
 
-  if(HAL_ADC_DeInit(&AdcHandle) != HAL_OK) {
+  if(HAL_ADC_DeInit(AdcHandle) != HAL_OK) {
     return 0;
   }
 
   return uhADCxConvertedValue;
 }
 
+/**
+  * @brief  This function will use DMA for ADC conversion
+  * @param  pin               : the pin to use
+  * @param  pData             : pointer to buffer where write data
+  * @param  lData             : Number of sample
+  * @param  callback          : Callback function to execute when conversion is done
+  * @param  functionParameter : Callback function parameter
+  * @retval 1 is success, 0 if fail.
+  */
+bool adc_read_value_dma(PinName pin, uint32_t *pData, uint32_t lData,
+                            void (*callback)(void *user_data), void *functionParameter)
+{
+  void *peripheral;
+  ADC_ChannelConfTypeDef sConfig;
+  ADCContext_t *ADCxContext = NULL;
+  ADC_HandleTypeDef *AdcHandle = NULL;
+
+  /* ### - 1 - Initialize ADC peripheral #################################### */
+  peripheral  = pinmap_peripheral(pin, PinMap_ADC);
+  ADCxContext = get_adc_context(peripheral);
+
+  if((peripheral == NP) || (ADCxContext == NULL)) {
+    return 0;
+  }
+
+  ADCxContext->current_pin  = pin;                                             /* Needed for HAL_ADC_MspInit     */
+  ADCxContext->cb           = callback;                                        /* Register the callback function */
+  ADCxContext->user_data    = functionParameter;
+  AdcHandle                 = &(ADCxContext->ADCHandle);
+  AdcHandle->Instance       = peripheral;
+
+  if (HAL_ADC_DeInit(AdcHandle) != HAL_OK)
+  {
+    return 0;
+  }
```

Comment:
I find it pretty bothersome that for each call to `adc_read_value_dma` the ADC gets deinitialized, then initialized, finally sampled, and then deinitialized again. Yes, this is amortized over `lData` samples, but still... How lkong does all this de-init/init/de-init take? Why not create an API that lets the user init things, and then call a very lightweight function to collect a batch of samples?